man/cvslog.1
.TH cvslog 1 "2006-07-29" "logilab devtools"

.SH NAME
cvslog \- CVS log analysis tool


.SH SYNOPSIS
.B cvslog 
.RI [command] [options] [-- [cvs options] [-- cvs log options]]
.br

.SH COMMANDS 
.B info
.br
.B stats
.br

.SH OPTIONS 
.B -h / --help
.br
.B -r / --rlog
.br
.B -d / --days
.br

.SH DESCRIPTION
Part of logilab's devtools.

.SH LICENCE
(c) 2004-2006 Logilab. General Public Licence.

.SH SEE ALSO
/usr/share/doc/python-devtools/

.SH AUTHOR
Logilab <contact@logilab.fr>




ChangeLog
ChangeLog for logilab.devtools
==============================

2006-08-03  -- 0.10.0 
    * fixed spelling of dependant/dependent
    * Move to new Python packaging policy
    * logilab.el: insert current year in copyright (#10779)
    * checkpackage: correctly ignore mercurial files (#14462)
    * cvslog: added man page (#8730)
    * preparedistrib: added retry option when step failed (#8862)
    * debianize: removed (#14473)
    * debian package refactor (#14476)
    * lgp: one program to rule them all (#14467)

2006-05-31  --  0.9.0
    * buildpackage: offers to test the packages with piuparts
    * preparedistrib: offers to copy the COPYING file from a set of
      known licenses 
    * buildpackage: fixed missing DEBUILDER environment variable bug
    * dos2unix: removal (unused, buggy, conflicts with debian package tofrodos)
    * debianize: fix the way to handle subpackage's __init__.py file (there
      is now a subpackage_master boolean property in the __pkginfo__.py which
      tells if a package is handling the __init__.py file, so now only one 
      subpackage should set this to True and the others should depends on 
      this package)
    * debianize: fix Uploader to Uploaders in control
    * pkginfo: fix debian handler detection
    * makedistrib: don't ask to tag package if an error occurs
    * vcslib: added mercurial support

	
2006-01-10  --  0.8.0
    * debianize:
    * updated to handle site-python installation with architecture
        independent package (implements #10119)
    * consider the optional debian_uploader variable in package
        information 
    * fix .examples files generation with debianize
    * fix templates: control.standard-version is now 3.6.2, updated FSF
        address in debian licenses files
    * fix package type detection in preparedistrib (fix #9917)
	

2005-07-29  --  0.7.0
    * fix dh -i/-a option handing in rules file generated by debianize
      (#9700)
    * no more empty lines annoying emacs in rules files generated by
      debianize  (#9710)
    * better use of .docs files / dh_installdocs and .examples files
      / dh_installexamples in debianize
    * better pbuilder support in buildeb (#1045)
    * tagpackage try to tag only distributed files (#674)
    * catch EOFError and KeyboardInterrupt when waiting for an input to
      avoid traceback
    * fix svn_agent.log_info message format


2005-07-19  --  0.6.7
    * fix call to pycoverage with missing arguments (close #9584)
    * makedistrib force manifest regeneration when building debian
      packages (close #9134)

	
2005-06-20  --  0.6.6
    * fix for apycot checkers (python2.2 compat + handle apycot >= 0.7)

	

2005-06-14  --  0.6.5
    * new options on checkpackage to select tests to apply
	
    * handle optional dependency on logilab-common for the -test package

    * fix svn tagging (both svn and cvs tags now include the
      project's name)

    * fix SGML catalog checking for registered DTDs with another language
      than EN



2005-02-25  --  0.6.4
    * fixed various bugs in lib.changelog
    * added write-permission check before checkin


	
2005-02-16  --  0.6.3
    * fix pycoverage to not include python file which are not in a
      subpackage (i.e. directory without __init__.py  file)

    * debianize: fix html doc detection, fix bug with py2.3 and optik

    * fix emacs startup file so that cover mode is automatically loaded at
      startup


	
2005-01-21  --  0.6.2
    * fix svn tag command

    * fix cvs edited command

    * added copyright information for coverage.py in the README and in
      the debian copyright files
	
    * add 2.4 generation by default to debianize

    * fix ChangeLog and debian/changelog parsers/writers to be less
      intrusive
	  
    * new column %missing in pycoverage reports

    * included a new emacs mode for coverage files

	
	  
2004-12-22  --  0.6.1
    * add missing import to vcslib.cvs (fix #8520)



2004-12-16  --  0.6.0
    * using optparse instead of getopt in dos2unix, added "replacement"
      and "listing" options

    * added the missing dos2unix script to the distribution

    * allow comments in DEPENDS, RECOMMENDS and SUGGEST files (using #
      at the begining of the comment line)

    * check_manifest and check_bin ignore .svn directory

    * removed the_coverage global in lib/coverage.py

    * fix max length and bullet lists in debian's descriptions

    * added missing end line in default debian/watch

    * major debianize refactoring using class / inheritance instead of
      functions composition

    * import the version control library from oobrother extended with code
      from devtools / apycot

    * support to read/write debian changelog file (see the --debian option
      of the changelog command)

    * consider lines beginning with a # as comment in DEPENDS, RECOMMENDS
      and SUGGESTS files



2004-11-03  --  0.5.4
    * fix __file__ problem with pycoverage

    * fixed -h option of pycoverage

    * checkpackage display output of executable which don't return 0 status
      when executed using -h or -help

    * tool 'dos2unix' added: removes '\r' inside specified files
      or all files of specified folders (which may be browsed
      recursively with option -r)



2004-10-18  --  0.5.3
    * licenses are now defined as text files in templates/licenses, so
      providing a way to recognize some new licenses

    * make buildeb, debianize and pycoverage more easily usable as a
      library

    * minor fix in changelog parsing

    * use new format in debian/watch

    * fix debianize's man pages handling



2004-07-07  --  0.5.2
    * fix cvs log parsing

    * fixed buildpackage with sdist target

    * buildeb may take a "command" argument, and use os.system instead of
      spawnv

    * pycoverage ignore "thirdparty" directory

    * spell check descriptions if snakespell is installed



2004-02-19  --  0.5.1
    * fix and enhance announce generation


cvslog.py
#!/usr/bin/env python
""" CVS Log analysis tool

USAGE: cvslog [COMMAND] [OPTIONS] [-- [CVS OPTIONS] [-- CVS LOG OPTION]]

COMMANDS:
    info
    stats

OPTIONS:
    -h / --help
    -r / --rlog
    -d / --days
"""

__revision__ = '$Id: cvslog.py,v 1.9 2004-02-13 09:21:32 syt Exp $'

import os
import sys
import getopt
from time import localtime, time, strftime
from logilab.devtools.vcslib.cvsparse import parse, LogLineHandler

def print_info(loginfo):
    for file in loginfo:
        print '*'*80
        print file
        for revision in file.children:
            print '-'*40
            print revision
            
def print_stats(loginfo, html=0, mincommit=None):
    if mincommit:
        mincommit = int(mincommit)

    results = {'authors': {}}
    
    for file in loginfo:
        for revision in file.children:
            author = revision.author
            results['authors'].setdefault(author, {'+': 0, '-': 0,
                                                   'new': 0, 'author': author,
                                                   'commit': 0, 'files': {}})
            d = results['authors'][author]
            d['files'][file.repo_file] = 1
            d['commit'] += 1
            if revision.initial:
                d['new'] += 1
            else:
                d['+'] += revision.lines[0]
                d['-'] -= revision.lines[1]

    nb_lines = 0
    winner = None
    max_commit = 0
    for author, data in results['authors'].items():
        if mincommit is not None and data['commit'] < mincommit:
            del results['authors'][author]
            continue
        if data['commit'] > max_commit:
            max_commit = data['commit']
            winner = author
        data['+commit'] = round(float(data['+'])/data['commit'])
        data['-commit'] = round(float(data['-'])/data['commit'])
        data['changed'] = data['+']-data['-']
        nb_lines += data['changed']
        data['changedcommit'] = round(float(data['changed'])/data['commit'])
        data['newcommit'] = round(float(data['new'])/data['commit'])
    results['winner'] = winner
    for author, data in results['authors'].items():
        data['%'] = round(float(data['changed']*100)/nb_lines)

    if html:
        from logilab.devtools.lib.cvsparse import LogResultHTMLPrinter
        LogResultHTMLPrinter().format(results)
    else:
        from logilab.devtools.lib.cvsparse import LogResultTextPrinter
        LogResultTextPrinter().format(results)
        
        
COMMANDS = {
    'info' : ( print_info,
               []
               ),
    'stats' : ( print_stats,
                ['date=', 'days=', 'html', 'min-commit=']
                )
    }

def get_logs(cvsoptions, cvscommandoptions, rlog=''):
    """extract an object log representation from a pipe with cvs
    """
    # construct cvs arguments
    if rlog:
        cvsargs = '%s rlog %s %s' % (' '.join(cvsoptions),
                                     ' '.join(cvscommandoptions), rlog)
    else:
        cvsargs = '%s log %s' % (' '.join(cvsoptions),
                                 ' '.join(cvscommandoptions))
    command = 'cvs -Q %s' % (cvsargs)
    # spawn the cvs command
    (p_input, p_output, p_err) = os.popen3(command)
    p_input.close()
    # and construct log info from its output
    phandler = LogLineHandler()
    parse(p_output, phandler)
    p_err.close()
    p_output.close()
    return phandler.root
    
def run(args):
    # split args
    i = 0
    myoptions = []
    cvsoptions = []
    cvscommandoptions = []
    order = (myoptions, cvsoptions, cvscommandoptions)
    for arg in args:
        if arg == '--':
            i += 1
        else:
            order[i].append(arg)
    # get command
    command = 'info'
    if myoptions and myoptions[0][0] != '-':
        command = myoptions.pop(0)
    command_handler, l_opt = COMMANDS[command]
    rlog = None
    options = {}
    date = None
    # parse my options
    l_opt = ['help', 'rlog=', 'days='] + l_opt
    opts, args = getopt.getopt(myoptions, 'hr:d:', l_opt)
    for opt, val in opts:
        if opt in ('-h','--help'):
            print __doc__
            return
        if opt in ('-r', '--rlog'):
            rlog = val
        elif opt in ('-d', '--days'):
            date = strftime("%Y/%m/%d", localtime(time()-int(val)*60*60*24))
            today = strftime("%Y/%m/%d", localtime(time()))
            cvscommandoptions.insert(0, '%s\<%s'%(date, today))
            cvscommandoptions.insert(0, '-d')
        else:
            options[opt.replace('-', '')] = val or 1
    loginfo = get_logs(cvsoptions, cvscommandoptions, rlog)
    # now do something with log info according to command
    command_handler(loginfo, **options)


if __name__ == '__main__':
    run(sys.argv[1:])
